### PR TITLE
TIP-652: rework import/export edit pages

### PR DESCRIPTION
**Pull request by juliensnz**
**Description (for Contributor and Core Developer)**

For the 1.7 version, we didn't have enough time to fully rework import/export screens and we decided to mix old twig and new pef architecture. It helped a lot to be able to ship on time but this shouldn't stay like that forever.

Problems:
- We cannot drop old twig/form/controller legacy stuff
- Hard to communicate between the two layers
- Not really nice to document

What have been done in this PR:
- Internal api to crud on job instance
- Internat api to launch a job (with a callback URL to the report)
- Form provider for job extension forms
- Transfer to new pef architecture
- 

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | :clock1:
| Added Behats                      | :white_check_mark:
| Changelog updated                 | :clock1:
| Review and 2 GTM                  |:red_circle:
| Micro Demo to the PO (Story only) | :negative_squared_cross_mark:
| Migration script                  | :red_circle:
| Tech Doc                          | :negative_squared_cross_mark:

:white_check_mark: Done and pass

:red_circle: Done but fail

:clock1: Done but pending

:negative_squared_cross_mark: Not needed

### CODE REVIEW

File: features/export/edit_an_export.feature [hunk 1]
```diff
@@ -68,7 +70,8 @@ Feature: Edit an export
       | Label | My export |
     Then I should see "There are unsaved changes."
 
-  @jira https://akeneo.atlassian.net/browse/PIM-5965
+  # TODO TIP-652
```
Comment (SamirBoulil) [inline features/export/edit_an_export.feature]: I leave a comment, just to make sure you noticed and ok with this.

File: features/export/product-export-builder/export_products_by_text.feature [hunk 2]
```diff
@@ -153,6 +153,7 @@ Feature: Export products according to text attribute filter
     Then I filter by "name" with operator "Contains" and value "Ranger"
     And I filter by "completeness" with operator "No condition on completeness" and value ""
     And I press "Save"
```
Comment (SamirBoulil) [inline features/export/product-export-builder/export_products_by_text.feature]: When I press "save", Then I should not see the text...

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php [hunk 3]
```diff
@@ -0,0 +1,345 @@
...
+        JobParametersFactory $jobParamsFactory,
+        JobLauncherInterface $simpleJobLauncher,
+        TokenStorageInterface $tokenStorage,
+        RouterInterface $router,
+        FormProviderInterface $formProvider
+    ) {
+        $this->repository            = $repository;
+        $this->jobRegistry           = $jobRegistry;
+        $this->jobInstanceNormalizer = $jobInstanceNormalizer;
+        $this->normalizer            = $normalizer;
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php]: Why do you need this normalizer for ?

File: src/Pim/Bundle/EnrichBundle/Resources/config/providers.yml [hunk 4]
```diff
@@ -57,6 +58,65 @@ services:
...
+                xlsx_currency_import: pim-job-instance-csv-base-import
+                xlsx_family_import: pim-job-instance-csv-base-import
+                xlsx_group_import: pim-job-instance-csv-base-import
+                xlsx_group_type_import: pim-job-instance-csv-base-import
+                xlsx_locale_import: pim-job-instance-csv-base-import
+                xlsx_product_import: pim-job-instance-csv-product-import
+                xlsx_variant_group_import: pim-job-instance-csv-base-import
+        tags:
+            - { name: pim_enrich.provider.form, priority: 100 }
+
```
Comment (fitn) [inline src/Pim/Bundle/EnrichBundle/Resources/config/providers.yml]: Each time, we will create a new job instance, we should override this service?

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/save.js [hunk 5]
```diff
@@ -0,0 +1,72 @@
+'use strict';
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/save.js]: move it to proper generic place

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/field/decimal-separator.js [hunk 6]
```diff
@@ -0,0 +1,27 @@
+'use strict';
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/field/decimal-separator.js]: add to generic location

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php (hunk 3 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php]: Yup, no need, I removed it

File: features/export/product-export-builder/export_products_by_text.feature (hunk 2 above)
Comment (juliensnz) [inline features/export/product-export-builder/export_products_by_text.feature]: fixed

File: src/Pim/Bundle/EnrichBundle/Resources/config/providers.yml (hunk 4 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/config/providers.yml]: No, it's a tagged service so you can add your with higher priority :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/field/date-format.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
...
+    return SelectField.extend({
+        /**
+         * {@inherit}
+         */
+        configure: function () {
+            return $.when(
+                FetcherRegistry.getFetcher('formats').fetchAll(),
+                SelectField.prototype.configure.apply(this, arguments)
+            ).then(function (formats) {
+                this.config.options = formats.date_formats;
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/field/date-format.js]: not really doable without adding some code

File: features/Pim/Behat/Context/NavigationContext.php [hunk 8]
```diff
@@ -360,13 +360,15 @@ public function iShouldSeeANiceLoadingMessage()
      *
      * @return \SensioLabs\Behat\PageObjectExtension\PageObject\Page
      */
-    public function openPage($pageName, array $options = [])
+    public function openPage($pageName, array $options = [], $wait = true)
```
Comment (juliensnz) [inline features/Pim/Behat/Context/NavigationContext.php]: done

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/content/data.js [hunk 9]
```diff
@@ -281,15 +283,21 @@ define(
              * Updates the form model by iterating over filter views
              */
             updateModel: function () {
+                var data = this.getFormData();
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/content/data.js]: needed

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/content/data.js [hunk 10]
```diff
@@ -281,15 +283,21 @@ define(
...
+                }
+
                 var dataFilterCollection = [];
 
                 _.each(this.filterViews, function (filterView) {
                     if (!filterView.isEmpty()) {
                         dataFilterCollection.push(filterView.getFormData());
                     }
                 });
+                data.configuration.filters.data = dataFilterCollection;
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/product/edit/content/data.js]: done

File: features/Pim/Behat/Context/Domain/Collect/ImportProfilesContext.php [hunk 11]
```diff
@@ -89,30 +89,16 @@ function ($row) use ($delimiter) {
...
-    public function iUploadAndImportTheFile($file)
+    public function iUploadAndImportTheFile($operator, $file)
     {
         $this->getMainContext()->getSubcontext('job')
             ->attachFileToField($this->replacePlaceholders($file), 'Drag and drop a file or click here');
+        $this->getCurrentPage()
+            ->getSession()
+            ->executeScript('$(\'.AknMediaField-fileUploaderInput\').trigger(\'change\');');
         $this->getCurrentPage()->pressButton('Upload and import now');
-
```
Comment (SamirBoulil) [inline features/Pim/Behat/Context/Domain/Collect/ImportProfilesContext.php]: 🔥

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php [hunk 12]
```diff
@@ -238,19 +341,12 @@ protected function getJobInstance($code, $checkStatus = true)
...
         if (null === $jobInstance) {
-            throw new NotFoundHttpException('Akeneo\Component\Batch\Model\JobInstance entity not found');
-        }
-
-        // Fixme: should look at the job execution to see the status of a job instance execution
-        if ($checkStatus && $jobInstance->getStatus() === JobInstance::STATUS_IN_PROGRESS) {
-            throw new NotFoundHttpException(
-                sprintf('The %s "%s" is currently in progress', $jobInstance->getType(), $jobInstance->getLabel())
-            );
+            throw new NotFoundHttpException(sprintf('%s entity not found', 'Akeneo\Component\Batch\Model\JobInstance'));
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php]: I think @skeleton was talking about adding a sprintf to put the `$code` of the job (for debug purpose) :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/product/updated.js [hunk 13]
```diff
@@ -129,7 +129,10 @@ define([
             if ('>' === operator) {
                 value = DateFormatter.format(value, DateContext.get('date').format, this.modelDateFormat);
             } else if ('SINCE LAST JOB' === operator) {
-                value = this.getParentForm().getFormData().jobCode;
+                value = this.getParentForm().getFormData().code;
+            }
+            if (undefined === value) {
```
Comment (SamirBoulil) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/product/updated.js]: _.isUndefined ? or typeof(value) === 'undefined' ?

File: src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml [hunk 14]
```diff
@@ -1,19 +1,39 @@
...
+    defaults: { _controller: pim_enrich.controller.rest.job_instance:deleteImportAction, _format: json }
     methods: [DELETE]
 
-pim_enrich_job_instance_rest_launch:
-    path: /rest/{code}/launch
-    defaults: { _controller: pim_enrich.controller.rest.job_instance:launchAction, _format: json }
+pim_enrich_job_instance_rest_import_launch:
+    path: /rest/import/{code}/launch
+    defaults: { _controller: pim_enrich.controller.rest.job_instance:launchImportAction, _format: json }
+    methods: [POST, GET]
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml]: Wasn't the GET supposed to be dropped for the launch action?

File: src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml [hunk 15]
```diff
@@ -1,19 +1,39 @@
...
+    defaults: { _controller: pim_enrich.controller.rest.job_instance:putExportAction, _format: json }
+    methods: [PUT, POST]
+
+pim_enrich_job_instance_rest_export_delete:
+    path: /rest/export/{code}
+    defaults: { _controller: pim_enrich.controller.rest.job_instance:deleteExportAction, _format: json }
+    methods: [DELETE]
+
+pim_enrich_job_instance_rest_export_launch:
+    path: /rest/export/{code}/launch
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml]: Why are they sometimes `{code}` and sometimes `{identifier}`, don't they refer to the same thing?

File: src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php [hunk 16]
```diff
@@ -136,11 +244,9 @@ public function getAction($identifier)
      * @param Request $request
      * @param string  $identifier
      *
-     * @AclAncestor("pim_importexport_export_profile_edit")
-     *
      * @return JsonResponse
      */
-    public function putAction(Request $request, $identifier)
+    protected function putAction(Request $request, $identifier)
```
Comment (solivier) [inline src/Pim/Bundle/EnrichBundle/Controller/Rest/JobInstanceController.php]: Add to BC

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/export/edit/save.js [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+    [
+        'pim/job-instance-edit-form/save',
+        'pim/saver/job-instance-export'
+    ],
+    function (
+        BaseSave,
+        JobInstanceSaver
+    ) {
+        return BaseSave.extend({
+            getJobInstanceSaver: function () {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/export/edit/save.js]: jsdoc

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/delete.js [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+/**
+ * Delete extension for job instance
+ *
+ * @author    Julien Sanchez <julien@akeneo.com>
+ * @copyright 2016 Akeneo SAS (http://www.akeneo.com)
+ * @license   http://opensource.org/licenses/osl-3.0.php  Open Software License (OSL 3.0)
+ */
+define(['pim/form/common/delete', 'pim/remover/job-instance-import'], function (DeleteForm, JobInstanceRemover) {
+    return DeleteForm.extend({
+        remover: JobInstanceRemover
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/delete.js]: jsdoc

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/save.js [hunk 19]
```diff
@@ -0,0 +1,25 @@
...
+    [
+        'pim/job-instance-edit-form/save',
+        'pim/saver/job-instance-import'
+    ],
+    function (
+        BaseSave,
+        JobInstanceSaver
+    ) {
+        return BaseSave.extend({
+            getJobInstanceSaver: function () {
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/save.js]: jsdoc

File: src/Pim/Bundle/EnrichBundle/Resources/views/JobInstance/form.html.twig [hunk 20]
```diff
@@ -25,27 +25,46 @@
...
 
                                     fetcherRegistry.initialize().done(function () {
-                                        fetcherRegistry.getFetcher('job_instance').fetch('{{ jobInstanceIdentifier }}').then(function (jobProfile) {
+                                        fetcherRegistry.getFetcher('job_instance_{{ type }}').fetch('{{ jobInstanceIdentifier }}').then(function (jobProfile) {
                                             formBuilder.build(jobProfile.meta.form + '-{{ mode }}')
                                                 .then(function (form) {
+                                                    var label      = _.escape(jobProfile.label);
+                                                    var newTitle = document.title.replace('%job.label%', label);
+
+                                                    document.title  = newTitle;
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/views/JobInstance/form.html.twig]: you already did a `document.title.replace()` above, I think you don't need to do an assignment here

File: src/Pim/Bundle/EnrichBundle/Resources/views/JobInstance/form.html.twig [hunk 21]
```diff
@@ -25,27 +25,46 @@
...
+                                                    var navigation  = Navigation.getInstance();
+                                                    var navTitle    = navigation.getPinButtonsData('title');
+                                                    navTitle.params = {'%job.label%': label};
+                                                    navigation.setPinButtonsData('title', navTitle);
+                                                    navigation.setPinButtonsData('title-rendered-short', newTitle);
+                                                    HistoryItemManager.save(navigation.url, navTitle);
+
+
+
```
Comment (grena) [inline src/Pim/Bundle/EnrichBundle/Resources/views/JobInstance/form.html.twig]: extra spaces 💃

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/common/edit/upload-launch.js [hunk 22]
```diff
@@ -0,0 +1,76 @@
...
+
+            /**
+             * Launch the job
+             */
+            launch: function () {
+                if (this.getFormData().file) {
+                    var formData = new FormData();
+                    formData.append('file', this.getFormData().file);
+
+                    $.ajax({
```
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/common/edit/upload-launch.js]: Not possible

File: src/Pim/Bundle/ImportExportBundle/Controller/JobProfileController.php [hunk 23]
```diff
@@ -202,356 +153,41 @@ public function createAction(Request $request)
...
-    {
-        $rawParameters = $jobInstance->getRawParameters();
-        $job = $this->jobRegistry->get($jobInstance->getJobName());
-        $jobParameters = $this->jobParametersFactory->create($job, $rawParameters);
-
-        /** @var ConstraintViolationListInterface $jobParamsViolations */
-        $jobParamsViolations = $this->jobParametersValidator->validate(
-            $job,
-            $jobParameters,
-            $validationGroups
```
Comment (juliensnz) [inline src/Pim/Bundle/ImportExportBundle/Controller/JobProfileController.php]: not relevant anymore

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/product/updated.js (hunk 13 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/filter/product/updated.js]: Why ?

File: src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml (hunk 15 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml]: explained on mob review. The responsible to this nightmare is the oro datagrid :)

File: src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/delete.js (hunk 18 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/public/js/job/import/edit/delete.js]: no need on variables

File: features/import/edit_an_import.feature [hunk 24]
```diff
@@ -35,18 +34,32 @@ Feature: Edit an import
...
-    And I should see "Escape \\"
-    And I should see "Real time history update No"
-    And I should see "Enable the product No"
-    And I should see "Categories column cat"
-    And I should see "Family column fam"
-    And I should see "Groups column grp"
-    And I should see "Decimal separator dot (.)"
-    And I should see "Date format yyyy-mm-dd"
+    And I should not see the text "There are unsaved changes."
+
```
Comment (juliensnz) [inline features/import/edit_an_import.feature]: extra line

File: src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml (hunk 14 above)
Comment (juliensnz) [inline src/Pim/Bundle/EnrichBundle/Resources/config/routing/job_instance.yml]: to remove indeed

**Comment by juliensnz** [comment]
Mob approved by bunch of people

**Pull request merged**